README.md
Studio 440 OS
=============

Source code for the Sequential Studio 440 operating system.


CHANGES


Running log of feature/function changes, bug fixes:


-------------------------------------------------------------------------------

2.13:

Sample dump, of course.

When creating sound (or loading 2000 sound from disk, or receive via SDS),
VCA release rate defaults to 10 instead of 40 -
mainly to ease sample start-point/end-point editing.

After sampling, sampled sound's kit is not automatically installed on pads.
(This was being done immediately upon sample return, and often caused
spurious changes to CURRENT_SOUND immediately following sampling, such that
new sample was usually lost.  Workaround - select sound for sampling using
pads, to avoid spurious change to different sample number).
See SET_ROOT_PITCH, SAMSUBS.

-------------------------------------------------------------------------------

2.14

Minor modifications to sample dump header interpretation/generation:
on transmit, set loop points = end if loop mode = off.
on receive, set loop points = end if beyond end (per Akai S1000, P3000) -
               also, set loop mode = off if loop points are equal.

-------------------------------------------------------------------------------

2.20

Same damned thing as 2.14, but formal release version.

-------------------------------------------------------------------------------

2.21

Modification of SCSI routines to accommodate standard CCS drives.
Until further notice, limited to 100 banks per drive (approx 80 meg).
Added user_id_screen display to power-up procedure,
and encoded all power-up screens for obscurity.

-------------------------------------------------------------------------------

2.22


09-May-90

Added resetjmp module, to permit standardized, constant entry address
upon power-up or other hard reset - used 100h as entry point.


21-Jun-90

space reclamation - begin with codecapp = 1f8de:
       .  removal of work loop function - codecapp = 1f4c8
       .  removal of drive_sel_sub - codecapp = 1f462
       .  removal of all code related to MTC clocking - codecapp = 1ec66


07-Aug-90

decode_scrnbuf moved from reset_44 into lcd as WRITE_ENCODED_SCREEN,
added to LCD_FUNS as WR_ENCD_SCRN.  now copies encoded text to screen
buffer and then decodes it, thus operates like WR_SCRN for encoded text.
also implemented DISP_ENCODED_SCREEN / DSP_ENCD_SCRN.

first-pass copy protection routines installed -
armed by user's first function switch hit, executed after timeout.


24-oct-90

change made in perf_mid to fix up a mapped-mode stuck-note bug -
refer to that file for details.
unprotected version given to peter rudolfi, who discovered the bug -
not sure if this included updated scsi stuff, or what became of the rom files.


09-oct-91

unprotected version custom-built for ken reyes - he got a syquest drive.
defeated copy-pro by removing backgrou branch to do_copypro_check.
just for the heck of it - did this by putting a conditional switch in equates.
don't forget that you did this!  maybe you'll put it back immediately ....
also updated stripinc.bat to reflect new environment.

-------------------------------------------------------------------------------

2.23


920813
restored absent smpte sync code from banished pro_mtc as pro_tcm.
a customized copy of this was shipped to luc allen.

921008
assorted changes to make the scsi implementation more robust
with respect to syquest/ccs drives.  eliminated user_id_screen
display as part of powerup init - this is for custom versions only.
reset procedure now explicitly initializes stack pointer,
for use with generic mon68k (hp64000 emulation monitor).

i called this version 2.30 - shipped it to dave seznak.


BASIC/ALEVLCAL.BAS
10       REM  PROGRAM TO COMPUTE VOLUME LEVEL TABLE VALUES.       
20       REM  RESULTS SENT TO SCREEN AND EITHER: COM1 OR FILE (VCALEVLS).
30       REM  RESULTS ARE GIVEN DIRECTLY IN 68000 SOURCE FORMAT.
40       REM  COMPUTES SCALE FACTORS FOR EACH OF VOLUME SETTINGS 1-31.
50       REM  SETTING 0 IS A SPECIAL CASE DONE W/O COMPUTATION.
60       REM  IN USE, SCALE FACTOR SCALES RAW VCA ENVELOPE AT EACH UPDATE TIME.
70      REM  ONLY 12 M.S.BITS OF VCA CONTROL OUTPUT WORD ARE SIGNIFICANT -
80      REM  HOWEVER, CONTROL VALUE AND SCALE FACTORS ARE MAINTAINED AT
90      REM  16-BITS TO PROVIDE SUFFICIENT LEVEL RESOLUTION.
100       PRINT "Transmit results over COM1?" 
110      INPUT " ( 0 = no, 1 = yes) ... : ",TRANS% 
120      IF TRANS% THEN OPEN "COM1:9600,N,8" AS #1 ELSE OPEN "O",#1,"VCALEVLS"
130      IF TRANS%=0 THEN PRINT "Results will be saved in file VCALEVLS -"
140      INPUT "Enter maximum attenuation (dB) - for level = 1: ",MAXDB#
150      STEPDB# = MAXDB#/30            
160     PRINT "             DC.W        0           ;  0    INF.    0.00000000"
170     PRINT #1,"             DC.W        0           ;  0    INF.    0.00000000"
180     FOR LEVELSETTING% = 1 TO 31
190     LEVELDB# = (31-LEVELSETTING%)*STEPDB#
200     LEVELLOG# = (LEVELDB#/10)*LOG(10)
210     LEVELFACTOR# = EXP(-LEVELLOG#)
220     SCALEDFACTOR# = LEVELFACTOR#*65535!
230     PRINT "             DC.W    ";
240     PRINT USING "#####           ";SCALEDFACTOR#;
250     PRINT "; ";
260     PRINT USING "##    ";LEVELSETTING%;
270     PRINT USING "##.#    ";LEVELDB#;
280     PRINT USING "#.########";LEVELFACTOR#
290     PRINT #1,"               DC.W    ";
300     PRINT #1,USING "#####           ";SCALEDFACTOR#;
310     PRINT #1,"; ";
320     PRINT #1,USING "##   ";LEVELSETTING%; 
330     PRINT #1,USING "##.#   ";LEVELDB#;
340     PRINT #1,USING "#.########";LEVELFACTOR#
350     NEXT LEVELSETTING% 
360     CLOSE #1
#1,USING "##.#   ";LEVELDB#;
340     PRINT #1,USING "#.########";LEVELFACTOR#
350     NEXT 

BASIC/ARATECAL.BAS
10       REM  PROGRAM TO COMPUTE VCA RATE FACTORS FOR EXPONENTIAL ENVELOPES.
20       REM  COMPUTES FACTORS FOR GROWING OR DECAYING ENVELOPES.
30       REM  RESULTS SENT TO SCREEN AND EITHER: COM1 OR FILE (VCARATES).
40       REM  RESULTS ARE GIVEN DIRECTLY IN 68000 SOURCE FORMAT.
50       REM  COMPUTES SCALING FACTORS FOR EACH OF RATE SETTINGS 1-98.
60       REM  SETTINGS 0 AND 99 ARE SPECIAL CASES DONE W/O COMPUTATION.
70       REM  ASSUMES ENVELOPE ENDPOINTS ARE 0FFFH AND 0001H (EITHER WAY).
80       REM  IN USE, VCA LEVEL IS SCALED BY RATE FACTOR AT EACH UPDATE TIME.
90      PRINT "Transmit results over COM1?" 
100      INPUT " ( 0 = no, 1 = yes) ... : ",TRANS% 
110      IF TRANS% THEN OPEN "COM1:9600,N,8" AS #1 ELSE OPEN "O",#1,"VCARATES"
120      IF TRANS%=0 THEN PRINT "Results will be saved in file VCARATES -"
130      INPUT "Enter the time for maximum attack/decay (seconds): ",MAXTIME#
140      MAXUPDATES# = MAXTIME#/8.000001E-03 
150      K# = (LOG(MAXUPDATES#)-LOG(2))/97
160      Z# = LOG(4096)
170      PRINT "Specify type of envelope curve -"
180      PRINT " 0 = decaying (negative) exponential"
190     PRINT " 1 = growing (positive) exponential" 
200     INPUT " -------->  : ",CURVETYPE
210     IF CURVETYPE<>1 THEN CURVETYPE = -1 
220     Z# = CURVETYPE*Z# 
230     PRINT "             DC.L    0000000000      ;  0    INF.   (NULL)    1.0000000000"
240     PRINT #1,"             DC.L    0000000000      ;  0    INF.   (NULL)    1.0000000000"
250     FOR RATESETTING% = 1 TO 98
260     NUMOFUPDATES# = 2*EXP(K#*(98-RATESETTING%))
270     FULLTIME# = NUMOFUPDATES#*8.000001E-03
280     UPDATEFACTOR# = EXP(Z#/NUMOFUPDATES#)
290     IF CURVETYPE=1 THEN GOTO 320
300     SCALEDFACTOR# = INT(UPDATEFACTOR#*65536#*65536#)  
310     GOTO 330
320     SCALEDFACTOR# = INT((UPDATEFACTOR#*65536#)+.5) 
330     PRINT "             DC.L    ";
340     PRINT USING "##########      ";SCALEDFACTOR#;
350     PRINT "; ";
360     PRINT USING "##   ";RATESETTING%; 
370     PRINT USING "##.##   ";FULLTIME#;
380     PRINT USING "####.#   ";NUMOFUPDATES#; 
390     PRINT USING "##.##########";UPDATEFACTOR#
400     PRINT #1,"               DC.L    ";
410     PRINT #1,USING "##########      ";SCALEDFACTOR#;
420     PRINT #1,"; ";
430     PRINT #1,USING "##   ";RATESETTING%; 
440     PRINT #1,USING "##.##   ";FULLTIME#;
450     PRINT #1,USING "####.#   ";NUMOFUPDATES#; 
460     PRINT #1,USING "##.##########";UPDATEFACTOR#
470     NEXT RATESETTING% 
480     CLOSE #1
###.#   ";NUMOFUPDATES#; 
4

BASIC/FCUTFCAL.BAS
10       REM  PROGRAM TO COMPUTE VCF CUTOFF VALUES.
20       REM  RESULTS SENT TO SCREEN AND EITHER: COM1 OR FILE (VCFCUTOF).
30       REM  RESULTS ARE GIVEN DIRECTLY IN 68000 SOURCE FORMAT.
40       REM  COMPUTES CONTROL VALUES FOR CUTOFF SETTINGS 0-99.
50       REM  VCF CONTROL RANGE IS 0-FFFH - VALUES ARE LEFT-JUST IN 16 BITS.
60       REM  VCF IS ASSUMED TO HAVE EXPONENTIAL CONTROL OF CUTOFF FREQUENCY.
70      PRINT "Transmit results over COM1?" 
80      INPUT " ( 0 = no, 1 = yes) ... : ",TRANS% 
90      IF TRANS% THEN OPEN "COM1:9600,N,8" AS #1 ELSE OPEN "O",#1,"VCFCUTOF"
100      IF TRANS%=0 THEN PRINT "Results will be saved in file VCFCUTOF -"
110     FOR CUTOFFSETTING% = 0 TO 99
120     CUTOFFVALUE# = (65535!/99)*CUTOFFSETTING%
130     EQUIVVALUE% = INT(CUTOFFVALUE#/16)
140     PRINT "             DC.W    ";
150     PRINT USING "#####           ";CUTOFFVALUE#;
160     PRINT "; ";
170     PRINT USING "##       ";CUTOFFSETTING%;
180     PRINT HEX$(EQUIVVALUE%)
190     PRINT #1,"             DC.W    ";
200     PRINT #1,USING "#####           ";CUTOFFVALUE#;
210     PRINT #1,"; ";
220     PRINT #1,USING "##       ";CUTOFFSETTING%;
230     PRINT #1,HEX$(EQUIVVALUE%)
240     NEXT CUTOFFSETTING%
250     CLOSE #1
PRINT #1,USING "##   

BASIC/FRATECAL.BAS
10       REM  PROGRAM TO COMPUTE VCF RATE CONSTANTS FOR ENVELOPES.
20       REM  RESULTS SENT TO SCREEN AND EITHER: COM1 OR FILE (VCFRATES).
30       REM  RESULTS ARE GIVEN DIRECTLY IN 68000 SOURCE FORMAT.
40       REM  COMPUTES DELTA CONSTANTS FOR EACH OF RATE SETTINGS 1-98.
50       REM  SETTINGS 0 AND 99 ARE SPECIAL CASES DONE W/O COMPUTATION.
60       REM  ASSUMES ENVELOPE ENDPOINTS ARE 0FFFH AND 0001H (EITHER WAY).
70       REM  IN USE, RATE CONSTANT ADDED TO VCF ENVELOPE AT EACH UPDATE TIME.
80      REM  ONLY 12 M.S.BITS OF VCF CONTROL OUTPUT WORD ARE SIGNIFICANT -
90      REM  HOWEVER, CONTROL VALUE AND DELTA CONSTANTS ARE MAINTAINED AT
100      REM  16-BITS TO PROVIDE SUFFICIENT RATE RESOLUTION.
110       PRINT "Transmit results over COM1?" 
120      INPUT " ( 0 = no, 1 = yes) ... : ",TRANS% 
130      IF TRANS% THEN OPEN "COM1:9600,N,8" AS #1 ELSE OPEN "O",#1,"VCFRATES"
140      IF TRANS%=0 THEN PRINT "Results will be saved in file VCFRATES -"
150      INPUT "Enter the time for maximum attack/decay (seconds): ",MAXTIME#
160      MAXUPDATES# = MAXTIME#/8.000001E-03 
170      K# = (LOG(MAXUPDATES#)-LOG(2))/97
180     PRINT "             DC.W        0           ;  0    INF.    INF.    0"
190     PRINT #1,"             DC.W        0           ;  0    INF.    INF.    0"
200     FOR RATESETTING% = 1 TO 98
210     NUMOFUPDATES# = 2*EXP(K#*(98-RATESETTING%))
220     FULLTIME# = NUMOFUPDATES#*8.000001E-03
230     UPDATEDELTA# = 65536!/NUMOFUPDATES#
240     SCALEDFACTOR# = INT(UPDATEDELTA#) 
250     PRINT "             DC.W    ";
260     PRINT USING "#####           ";SCALEDFACTOR#;
270     PRINT "; ";
280     PRINT USING "##   ";RATESETTING%; 
290     PRINT USING "##.##   ";FULLTIME#;
300     PRINT USING "####.#   ";NUMOFUPDATES#; 
310     PRINT HEX$(SCALEDFACTOR#/16)
320     PRINT #1,"               DC.W    ";
330     PRINT #1,USING "#####           ";SCALEDFACTOR#;
340     PRINT #1,"; ";
350     PRINT #1,USING "##   ";RATESETTING%; 
360     PRINT #1,USING "##.##   ";FULLTIME#;
370     PRINT #1,USING "####.#   ";NUMOFUPDATES#; 
380     PRINT #1,HEX$(SCALEDFACTOR#/16)
390     NEXT RATESETTING% 
400     PRINT "             DC.W    65535           ; 99    0.00      1.0   FFF"
410     PRINT #1,"             DC.W    65535           ; 99    0.00      1.0   FFF"
420     CLOSE #1
  0.00      1.0   FFF"
410     PRINT #1,"             DC.W    65535        

BASIC/FTIMECAL.BAS
10       REM  PROGRAM TO COMPUTE VCF RATE DELTAS FOR ATTACK/DECAY ENVELOPES.
20       REM  BASICALLY SAME PROGRAM AS FRATECAL.BAS, REARRANGED SO THAT
30       REM  HIGHER SETTINGS CORRESPOND TO LONGER TIMES (SLOWER RATES?)
40       REM  COMPUTES FACTORS FOR LINEAR ENVELOPES - ONLY ONE TYPE USEABLE,
50       REM  SINCE VCF INPUT RESPONDS EXPONENTIALLY TO CONTROL VOLTAGE.
60       REM  RESULTS SENT TO SCREEN AND EITHER: COM1 OR FILE (VCFTIMES).
70       REM  RESULTS ARE GIVEN DIRECTLY IN 68000 SOURCE FORMAT.
80       REM  COMPUTES RATE DELTAS FOR EACH OF RATE SETTINGS 1-98.
90       REM  YAH, WE STILL CALL 'EM RATES, EVEN THOUGH THEY'RE REALLY TIMES.
100       REM  AH WELL - WHAT THE FUCK CAN YA DO ....
110       REM  SETTINGS 0 AND 99 ARE SPECIAL CASES DONE W/O COMPUTATION.
120       REM  PREVIOUSLY ASSUMED ENVELOPE ENDPOINTS 0FFFH AND 0001H -
130       REM  NOW COMPUTES ON BASIS OF ENVELOPE ENDPOINTS 0FFFH AND 0030H,
140       REM  SINCE VCF BELOW 30H TENDS TO BE INAUDIBLE IN THE 440.
150       REM  ALSO PREVIOUSLY COMPUTED EXPONENTIAL DISTRIBUTION OF TIMES
160       REM  FOR EACH OF THE SETTINGS - NOW USES PARABOLIC DISTRIBUTION,
170       REM  ALLOWING CONTROL OVER THE SHAPE OF THE DISTRIBUTION CURVE.
180       REM  IN 440, RATE DELTA IS ADDED TO / SUBTRACTED FROM VCF LEVEL
190      REM  AT EACH UPDATE TIME.
200      REM
210      PRINT "Transmit results over COM1?" 
220      INPUT " ( 0 = no, 1 = yes) ... : ",TRANS% 
230      IF TRANS% THEN OPEN "COM1:9600,N,8" AS #1 ELSE OPEN "O",#1,"VCFTIMES"
240      IF TRANS%=0 THEN PRINT "Results will be saved in file VCFTIMES -"
250      INPUT "Enter the time for maximum attack/decay (seconds): ",MAXTIME#
260      MAXUPDATES# = MAXTIME#/8.000001E-03 
270      REM
280      REM      K# = (LOG(MAXUPDATES#)-LOG(2))/97
290      REM      (this was part of exponential distribution algorithm).
300      REM
310      REM   (following is replacement parabolic exponent constant:)
320      PRINT "Specify nonlinearity of time distribution curve"
330      INPUT "(default = 3.000): ",K#
340      IF K# = 0 THEN K# = 3!
350      REM
360      PRINT ";"
370      PRINT "; (table computed using nonlinearity coefficient ";
380      PRINT USING "#.###";K#;
390      PRINT ")"
400      PRINT ";"
410      PRINT #1,";"
420      PRINT #1,"; (table computed using nonlinearity coefficient ";
430      PRINT #1,USING "#.###";K#;
440      PRINT #1,")"
450      PRINT #1,";"
460      REM
470      REM  insert zero-setting entry for rate-delta table:
480     PRINT "             DC.W    65535           ;  0     0.00      1.0  FFF"
490     PRINT #1,"               DC.W    65535           ;  0     0.00      1.0  FFF"
500      REM
510     FOR TIMESETTING% = 1 TO 98
520      REM
530      REM  (following was other part of exponential algorithm:)
540      REM  NUMOFUPDATES# = 2*EXP(K#*(TIMESETTING%-1))
550      REM  (whereas this is the parabolic version:)
560      NUMOFUPDATES# = 2 + ((TIMESETTING%/98)^K#)*(MAXUPDATES#-2)
570      REM
580     FULLTIME# = NUMOFUPDATES#*8.000001E-03
590      REM  this is the 0FFFH - 0030H part right here:
600     UPDATEDELTA# = (65535!-48!)/NUMOFUPDATES#
610     SCALEDFACTOR# = INT(UPDATEDELTA#) 
620     PRINT "             DC.W    ";
630     PRINT USING "#####           ";SCALEDFACTOR#;
640     PRINT "; ";
650     PRINT USING "##   ";TIMESETTING%; 
660     PRINT USING "###.##  ";FULLTIME#;
670     PRINT USING "#####.#  ";NUMOFUPDATES#; 
680     PRINT HEX$(SCALEDFACTOR#/16)
690     PRINT #1,"               DC.W    ";
700     PRINT #1,USING "#####           ";SCALEDFACTOR#;
710     PRINT #1,"; ";
720     PRINT #1,USING "##   ";TIMESETTING%; 
730     PRINT #1,USING "###.##  ";FULLTIME#;
740     PRINT #1,USING "#####.#  ";NUMOFUPDATES#; 
750     PRINT #1,HEX$(SCALEDFACTOR#/16)
760     NEXT TIMESETTING% 
770      REM
780      REM  insert max-setting entry for rate-delta table:
790     PRINT "             DC.W        0           ; 99     INF.     INF.  0"
800     PRINT #1,"               DC.W        0           ; 99     INF.     INF.  0"
810      REM
820     CLOSE #1


BASIC/STIMECAL.BAS
10       REM  PROGRAM TO COMPUTE VCA/VCF SUSTAIN TIME VALUES.
20       REM  BASICALLY SAME PROGRAM AS FTIMECAL.BAS - PROVIDES OUTPUT 
30       REM  EQUIVALENT TO THE "# OF UPDATES" COLUMN IN COMMENTS FIELD.
40       REM  RESULTS SENT TO SCREEN AND EITHER: COM1 OR FILE (SUSTIMES).
50       REM  RESULTS ARE GIVEN DIRECTLY IN 68000 SOURCE FORMAT.
60       REM  COMPUTES SUSTAIN TIMES FOR EACH OF RATE SETTINGS 1-99.
70        REM  SETTING 0 VALUE IS GENERATED WITHOUT COMPUTATION.
80       REM  SETTING -1 (GATE) IS HANDLED WITHOUT REFERENCE TO TABLE AT ALL.
90       REM  USES PARABOLIC DISTRIBUTION OF SUSTAIN TIMES,
100       REM  ALLOWING CONTROL OVER THE SHAPE OF THE DISTRIBUTION CURVE.
110       REM  IN 440, SUSTAIN TIME COUNT IS DECREMENTED AT EACH UPDATE TIME,
120       REM  SUSTAIN ENDS WHEN COUNT UNDERFLOWS.
130      REM
140      PRINT "Transmit results over COM1?" 
150      INPUT " ( 0 = no, 1 = yes) ... : ",TRANS% 
160      IF TRANS% THEN OPEN "COM1:9600,N,8" AS #1 ELSE OPEN "O",#1,"SUSTIMES"
170      IF TRANS%=0 THEN PRINT "Results will be saved in file SUSTIMES -"
180      INPUT "Enter the time for maximum sustain (seconds): ",MAXTIME#
190      MAXUPDATES# = MAXTIME#/8.000001E-03 
200      REM
210      REM   parabolic exponent constant:
220      PRINT "Specify nonlinearity of time distribution curve"
230      INPUT "(default = 3.000): ",K#
240      IF K# = 0 THEN K# = 3!
250      REM
260      PRINT ";"
270      PRINT "; (table computed using nonlinearity coefficient ";
280      PRINT USING "#.###";K#;
290      PRINT ")"
300      PRINT ";"
310      PRINT #1,";"
320      PRINT #1,"; (table computed using nonlinearity coefficient ";
330      PRINT #1,USING "#.###";K#;
340      PRINT #1,")"
350      PRINT #1,";"
360      REM
370      REM  insert zero-setting entry for sustain-time table:
380     PRINT "             DC.W        0           ;  0     0.00"
390     PRINT #1,"               DC.W        0           ;  0     0.00"
400      REM
410     FOR TIMESETTING% = 1 TO 99
420      REM  this is the parabolic distribution algorithm:
430      NUMOFUPDATES# = 1 + ((TIMESETTING%/99)^K#)*(MAXUPDATES#-1)
440      REM
450     FULLTIME# = NUMOFUPDATES#*8.000001E-03
460     PRINT "             DC.W    ";
470     PRINT USING "#####           ";NUMOFUPDATES#;
480     PRINT "; ";
490     PRINT USING "##   ";TIMESETTING%; 
500     PRINT USING "###.##  ";FULLTIME#
510     PRINT #1,"               DC.W    ";
520     PRINT #1,USING "#####           ";NUMOFUPDATES#;
530     PRINT #1,"; ";
540     PRINT #1,USING "##   ";TIMESETTING%; 
550     PRINT #1,USING "###.##  ";FULLTIME#
560     NEXT TIMESETTING%
570     CLOSE #1


BASIC/VTABLCAL.BAS
10       REM  PROGRAM TO COMPUTE PAD VELOCITY LOOKUP TABLES.
20       REM  RESULTS SENT TO SCREEN AND EITHER: COM1 OR FILE (VELTABLE).
30       REM  RESULTS ARE GIVEN DIRECTLY IN 68000 SOURCE FORMAT.
40       REM  COMPUTES SCALING FACTORS FOR EACH OF VELOCITIES 1-127,  
50       REM  ASSUMING RAW PAD LEVEL SHIFTED RIGHT ONE BIT, AND NO ZERO RESULT.
60      PRINT "Transmit results over COM1?" 
70      INPUT " ( 0 = no, 1 = yes) ... : ",TRANS% 
80      IF TRANS% THEN OPEN "COM1:9600,N,8" AS #1 ELSE OPEN "O",#1,"VELTABLE"
90      IF TRANS%=0 THEN PRINT "Results will be saved in file VELTABLE -"
100      PRINT
110      INPUT "Enter the curve exponent -  1.00000 = linear, etc. : ",CURVEXP#
120     PRINT ";"
130     PRINT "; (following table generated using exponent = ";
140     PRINT USING "##.###";CURVEXP#;
150     PRINT ")"
160     PRINT ";"
170     PRINT #1,";"
180     PRINT #1,"; (following table generated using exponent = ";
190     PRINT #1,USING "##.###";CURVEXP#;
200     PRINT #1,")"
210     PRINT #1,";"
220      REM 
230      REM  loop begins here:
240      REM
250     FOR VELOCITY = 1 TO 127
260     REM
270     REM do the comput here:
280     INTERNALVEL# = ((VELOCITY/127)^CURVEXP#)*31
290     IF INT(INTERNALVEL#) = 0 THEN INTERNALVEL# = 1
300     REM
310     REM output to screen:
320     PRINT "             DC.B    ";
330     PRINT USING "###             ";INTERNALVEL#;
340     PRINT "; ";
350     PRINT USING "###";VELOCITY
360     REM output to com1 or file:
370     PRINT #1,"               DC.B    ";
380     PRINT #1,USING "###             ";INTERNALVEL#;
390     PRINT #1,"; ";
400     PRINT #1,USING "###";VELOCITY
410     NEXT VELOCITY
420     CLOSE #1


BASIC/ATIMECAL.BAS
10       REM  PROGRAM TO COMPUTE VCA RATE FACTORS FOR EXPONENTIAL ENVELOPES.
20       REM  BASICALLY SAME PROGRAM AS ARATECAL.BAS, REARRANGED SO THAT
30       REM  HIGHER SETTINGS CORRESPOND TO LONGER TIMES (SLOWER RATES?)
40       REM  COMPUTES FACTORS FOR GROWING OR DECAYING EXPONENTIAL ENVELOPES.
50       REM  RESULTS SENT TO SCREEN AND EITHER: COM1 OR FILE (VCATIMES).
60       REM  RESULTS ARE GIVEN DIRECTLY IN 68000 SOURCE FORMAT.
70       REM  COMPUTES SCALING FACTORS FOR EACH OF RATE SETTINGS 1-98.
80       REM  YAH, WE STILL CALL 'EM RATES, EVEN THOUGH THEY'RE REALLY TIMES.
90       REM  AH WELL - WHAT THE FUCK CAN YA DO ....
100       REM  SETTINGS 0 AND 99 ARE SPECIAL CASES DONE W/O COMPUTATION.
110       REM  PREVIOUSLY ASSUMED ENVELOPE ENDPOINTS 0FFFH AND 0001H -
120       REM  NOW COMPUTES ON BASIS OF ENVELOPE ENDPOINTS 0FFFH AND 0030H,
130       REM  SINCE VCA BELOW 30H TENDS TO BE INAUDIBLE IN THE 440.
140       REM  ALSO PREVIOUSLY COMPUTED EXPONENTIAL DISTRIBUTION OF TIMES
150       REM  FOR EACH OF THE SETTINGS - NOW USES PARABOLIC DISTRIBUTION,
160       REM  ALLOWING CONTROL OVER THE SHAPE OF THE DISTRIBUTION CURVE.
170       REM  IN 440, VCA LEVEL IS SCALED BY RATE FACTOR AT EACH UPDATE TIME.
180      PRINT "Transmit results over COM1?" 
190      INPUT " ( 0 = no, 1 = yes) ... : ",TRANS% 
200      IF TRANS% THEN OPEN "COM1:9600,N,8" AS #1 ELSE OPEN "O",#1,"VCATIMES"
210      IF TRANS%=0 THEN PRINT "Results will be saved in file VCATIMES -"
220      INPUT "Enter the time for maximum attack/decay (seconds): ",MAXTIME#
230      MAXUPDATES# = MAXTIME#/8.000001E-03 
240      REM
250      REM      K# = (LOG(MAXUPDATES#)-LOG(2))/97
260      REM      (this was part of exponential distribution algorithm).
270      REM
280      REM   (following is replacement parabolic exponent constant:)
290      PRINT "Specify nonlinearity of time distribution curve"
300      INPUT "(default = 3.000): ",K#
310      IF K# = 0 THEN K# = 3!
320      REM
330      REM  (Z# embodies ratio of envelope endpoints:)
340      REM  (following was used for 0FFFH - 0001H endpoints:)
350      REM  Z# = LOG(4096)
360      REM  (following is used for 0FFFH - 0030H endpoints:)
370      Z# = LOG(4096/48)
380      REM
390      PRINT "Specify type of envelope curve -"
400      PRINT " 0 = decaying (negative) exponential"
410     PRINT " 1 = growing (positive) exponential" 
420     INPUT " -------->  : ",CURVETYPE
430     IF CURVETYPE<>1 THEN CURVETYPE = -1 
440     Z# = CURVETYPE*Z# 
450      PRINT ";"
460      PRINT "; (table computed using nonlinearity coefficient ";
470      PRINT USING "#.###";K#;
480      PRINT ")"
490      PRINT ";"
500      PRINT #1,";"
510      PRINT #1,"; (table computed using nonlinearity coefficient ";
520      PRINT #1,USING "#.###";K#;
530      PRINT #1,")"
540      PRINT #1,";"
550      REM
560      IF CURVETYPE = 1 THEN GOTO 620
570      REM  insert zero-setting entry for exponential-decay table:
580     PRINT "             DC.L    0000000000      ;  0    0.00       1.0    0.0000000000"
590     PRINT #1,"               DC.L    0000000000      ;  0    0.00       1.0    0.0000000000"
600      GOTO 660
610      REM
620      REM  insert zero-setting entry for exponential-growth table:
630      PRINT "             DC.L    0FFFFFFFFH      ;  0    0.00       1.0    65535.999999"
640      PRINT #1,"               DC.L    0FFFFFFFFH      ;  0    0.00       1.0    65535.999999"
650      REM
660     FOR TIMESETTING% = 1 TO 98
670      REM
680      REM  (following was other part of exponential algorithm:)
690      REM  NUMOFUPDATES# = 2*EXP(K#*(TIMESETTING%-1))
700      REM  (whereas this is the parabolic version:)
710      NUMOFUPDATES# = 2 + ((TIMESETTING%/98)^K#)*(MAXUPDATES#-2)
720      REM
730     FULLTIME# = NUMOFUPDATES#*8.000001E-03
740     UPDATEFACTOR# = EXP(Z#/NUMOFUPDATES#)
750     IF CURVETYPE=1 THEN GOTO 780
760     SCALEDFACTOR# = INT(UPDATEFACTOR#*65536#*65536#)  
770     GOTO 790
780     SCALEDFACTOR# = INT((UPDATEFACTOR#*65536#)+.5) 
790     PRINT "             DC.L    ";
800     PRINT USING "##########      ";SCALEDFACTOR#;
810     PRINT "; ";
820     PRINT USING "##  ";TIMESETTING%; 
830     PRINT USING "###.##   ";FULLTIME#;
840     PRINT USING "#####.#   ";NUMOFUPDATES#; 
850     PRINT USING "##.##########";UPDATEFACTOR#
860     PRINT #1,"               DC.L    ";
870     PRINT #1,USING "##########      ";SCALEDFACTOR#;
880     PRINT #1,"; ";
890     PRINT #1,USING "##  ";TIMESETTING%; 
900     PRINT #1,USING "###.##   ";FULLTIME#;
910     PRINT #1,USING "#####.#   ";NUMOFUPDATES#; 
920     PRINT #1,USING "##.##########";UPDATEFACTOR#
930     NEXT TIMESETTING% 
940      REM
950      IF CURVETYPE = 1 THEN GOTO 1010
960      REM  insert max-setting entry for exponential-decay table:
970      PRINT "             DC.L    0FFFFFFFFH      ; 99   ~INF.  ??(HUGE)    0.99999999999999999999999999....."
980      PRINT #1,"               DC.L    0FFFFFFFFH      ; 99   ~INF.  ??(HUGE)    0.99999999999999999999999999....."
990      GOTO 1050
1000      REM
1010      REM  insert max-setting entry for exponential-growth table:
1020      PRINT "             DC.L         65536      ; 99    INF.     INF.     1.0000000000"
1030      PRINT #1,"               DC.L         65536      ; 99    INF.     INF.     1.0000000000"
1040      REM
1050     CLOSE #1
